### DISCUSSION TITLE
Eglot capabilities are slower for me when display-line-numbers-mode is enabled. What could be the reason for this ?

### DISCUSSION THREAD

**Discussion by r-josef25** [joaotavora/eglot#1138] [Q&A]
Emacs Version: GNU Emacs 29.0.50 (build 2, x86_64-pc-linux-gnu, GTK+ Version 3.24.34, cairo version 1.17.6) of 2022-11-26

Steps to reproduce:
[(rust toolchain, rust-analyzer) and package: rust-mode should be available to reproduce this issue following the below given steps. Ideally, this should be reproducible on any setup for any language as long as the completion list to be displayed is considerably large.]

1) Start Emacs with emacs -Q.
2) load rust-mode and open up any cargo project(or you could create a new one).
3) enable eglot.
4) try import something from std library like:
    use::std::
   (a) Now run completion-at-point. 
   (b) Enable display-line-numbers-mode and then run completion-at-point again. 

There should be a noticeable difference in how long it takes completion windows to pop up for (a) and (b). [(b) will have higher latency than (a)]. If you do not notice any difference, try completing for something which could fetch longer completion list.

**Comment by joaotavora** [comment]
Hi @r-josef25 

Thanks for the simple reproduction recipe.  I've streamlined it in some shell commands

In the normal emacs session, make sure `rust-mode` is installed from {GNU |M}ELPA

```
$ cd /tmp
$ cargo new thingy
$ /path/to/recent/emacs -Q -f package-initialize /tmp/thingy/src/main.rs -f eglot
```

* Insert `use std::`
* `C-o` to open a newline
* `C-M-i` to request completions

This provides me with 114 candidates and the latency is small that I don't really notice it.

* `M-x display-line-numbers-mode`
* `C-M-i`

No discernible difference!

**Reply by r-josef25** [reply]
It could be because both completions are listed extremely quickly on your setup. In that case, profiling the two cases could tell if there is a difference. Unfortunately, I got no where with attempting profiling. If you have a fairly sized rust project on your system, you could try fetching completion list while within that project (try completion on something from the std library, something you know is going to produce a long completion list). I found latency difference to be more evident on my setup when I attempt completion or eldoc from within a project.

**Reply by joaotavora** [reply]
But can you do exactly what I did and confirm if you see a slowdown.  That could rule out a number of scenarios, like your observation being specific to a particular version of Emacs.

Even in big C++ I use projects, I don't have very long completion lists because the LSP servers cap it off at about 100 or so.

**Reply by r-josef25** [reply]
Following your steps, it seems that there is a slight difference. It is hard to say for sure. This issue is more prevalent for me when within a project. Its somehow worse when I start emacs with my init file.

Also, something I noticed while trying to reproduce the error through your steps is that the latency difference seems to be somehow change with dpi/scale factor of my compositor. Please let me know if you have any thoughts on this.

**Reply by joaotavora** [reply]
My thoughts is that you are experiencing something purely related to display-line-numbers-mode, and you connect it to Eglot because when using it you incidentally trigger a rearrangement  of windows and that's where your particular setup exhibits problems that i can't reproduce. My advice is to report this as an Emacs bug by sending email to bug-gnu-emacs@gnu.org.
